pkg/services/annotations/annotationsimpl/loki/alert_state_history_store.go

```diff
@@ -0,0 +1,399 @@
+		}
+
+		now := time.Now().UTC()
+		if query.To == 0 {
+			query.To = now.UnixMilli()
+		}
+		if query.From == 0 {
+			query.From = now.Add(-defaultQueryRange).UnixMilli()
+		}
+		if query.From > query.To {
+			to := query.To
+			query.To = query.From
+			query.From = to
+		}
+
+		res, err := r.client.RangeQuery(ctx, logQL, query.From, query.To, query.Limit)
+		if err != nil {
+			return ErrLokiStoreInternal.Errorf("failed to query loki: %w", err)
+		}
+
+		items = r.itemsFromStreams(ctx, query.OrgID, res.Data.Result, accessResources)
+
+		// order by time desc
+		sort.Slice(items, func(i, j int) bool {
+			return items[i].Time > items[j].Time
+		})
+
+		return nil
+	}()
+
+	return items, err
+}
+
+// itemsFromStreams builds annotation items from a list of Loki streams.
+// It will filter out items that the user does not have access to
+// and will skip items that cannot be parsed.
+func (r *AlertStateHistoryStore) itemsFromStreams(
+	ctx context.Context,
+	orgID int64,
+	streams []historian.Stream,
+	resources accesscontrol.AccessResources,
+) []*annotations.ItemDTO {
+	totalLen := 0
+	for _, stream := range streams {
+		totalLen += len(stream.Values)
+	}
+
+	rules := make(map[string]*ngmodels.AlertRule)
+	items := make([]*annotations.ItemDTO, 0, totalLen)
+
+	for _, stream := range streams {
+		for _, sample := range stream.Values {
+			var entry historian.LokiEntry
+			err := json.Unmarshal([]byte(sample.V), &entry)
+			if err != nil {
+				r.log.Debug("could not parse Loki entry", "err", err)
+				continue
+			}
+
+			transition, err := buildTransitionStub(&entry, sample.T)
+			if err != nil {
+				r.log.Debug("could not build state transition", "err", err)
+				continue
+			}
+
+			if !shouldReplay(entry, transition, resources) {
+				continue
+			}
+
+			rule, ok := rules[entry.RuleUID]
+			if !ok {
+				ruleQ := ruleQuery{OrgID: orgID, UID: entry.RuleUID}
+				rule, err = r.getRule(ctx, ruleQ)
+				if err != nil {
+					r.log.Debug(fmt.Sprintf("could not find rule with UID `%s`", entry.RuleUID), "entry", entry, "err", err)
+					continue
+				}
+
+				rules[entry.RuleUID] = rule
+			}
+
+			item, err := buildAnnotationItem(&entry, resources.Dashboards[entry.DashboardUID], rule, transition.State)
+			if err != nil {
+				r.log.Debug("could not build annotation item", "entry", entry, "err", err)
+				continue
+			}
+
+			items = append(items, item)
+		}
+	}
+	return items
+}
+
+type ruleQuery struct {
+	OrgID int64
+	UID   string
+	ID    int64
+}
+
+func (r *AlertStateHistoryStore) getRule(ctx context.Context, query ruleQuery) (*ngmodels.AlertRule, error) {
+	rule := &ngmodels.AlertRule{OrgID: query.OrgID}
+	if query.UID != "" {
+		rule.UID = query.UID
+	}
+	if query.ID != 0 {
+		rule.ID = query.ID
+	}
+
+	err := r.db.WithDbSession(ctx, func(sess *db.Session) error {
+		exists, err := sess.Get(rule)
+		if err != nil {
+			return err
+		}
+		if !exists {
+			return ErrMissingRule
+		}
+		return nil
+	})
+
+	return rule, err
+}
```

Comment:
Is there data we need from the original rule, that isn't already in Loki? Looking this up on read could lead to strange behavior - ASH, including annotations, is frozen in time even if the actual rule got updated or deleted later.

If there is data that we need from the Grafana database that's not available in Loki, but also is needed by annotations, IMO we should first start recording that data to Loki rather than fetch it from the SQL database.

Reply:
I wanted to avoid changing the ASH implementation if possible, but that's a very good point. It's easy enough to just add the rule title to what we're recording in Loki. 

We'll likely still need to do this lookup to translate ruleID to ruleUID, but afaik that should be stable on update and we shouldn't be querying Loki through annotations for a deleted rule.